src/vs/workbench/contrib/terminal/browser/media/shellIntegration.ps1

```diff
@@ -14,12 +14,23 @@ if ($ExecutionContext.SessionState.LanguageMode -ne "FullLanguage") {
 }
 
 $Global:__VSCodeOriginalPrompt = $function:Prompt
-
 $Global:__LastHistoryId = -1
-function Set-Serialized {
-	param ([string] $toSerialize)
-	$toSerialize = $toSerialize.Replace("\", "\\").Replace("`n", "\x0a").Replace(";", "\x3b")
-	return $toSerialize
+
+function Global:__VSCode-Escape-Value([string]$value) {
+	# NOTE: In PowerShell v6.1+, this can be written `$value -replace '…', { … }` instead of `[regex]::Replace`.
+	# Replace any non-alphanumeric characters.
+	[regex]::Replace($value, '[^a-zA-Z0-9]', { param($match)
+		# Backslashes must be doubled.
+		if ($match.Value -eq '\') {
+			'\\'
+		} else {
+			# Any other characters are encoded as their UTF-8 hex values.
+			-Join (
+				[System.Text.Encoding]::UTF8.GetBytes($match.Value)
+				| ForEach-Object { param($byte) '\x{0:x2}' -f $byte }
+			)
```

Comment:
Any ideas on the problem in the screenshots @rwe? This is testing my pwsh knowledge limits 🙂

Reply:
Looks like I dumbed something in that last commit by naming the parameter…PowerShell isn't my forte either! Thanks for checking.

This gives the expected encoding, I'll push it up in a sec:

```diff
-				| ForEach-Object { param($byte) '\x{0:x2}' -f $byte }
+				| ForEach-Object { '\x{0:x2}' -f $_ }
```

Also note that without #165635, the editor itself incorrectly decodes messages, as described there.

Reply:
Nice 👍. Yep, hoping to get to that one today. Thanks again for all the effort!

Reply:
Ok, I hope rebasing on `main` is OK — that line is the only change from the merge you'd resolved. (Modulo a single blank line). Pushing up now.